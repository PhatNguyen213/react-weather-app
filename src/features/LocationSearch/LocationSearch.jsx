import AutoSuggestDropdown from '../shared/AutoSuggestDropdown';
import { styled } from '@mui/material/styles';

const Div = styled('div')(({ theme }) => ({
  ...theme.typography.button,
}));

const LocationSearch = ({ suggestions, searchTerm, onChange, onSelect }) => {
  return (
    <>
      <Div>Please enter city name below for suggestions</Div>
      <AutoSuggestDropdown
        onSelect={onSelect}
        suggestions={suggestions}
        inputValue={searchTerm}
        onChange={onChange}
      />
    </>
  );
};

export default LocationSearch;
