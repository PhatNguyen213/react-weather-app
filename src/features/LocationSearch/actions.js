import { createAsyncThunk } from '@reduxjs/toolkit';
import { getLocationsFromText } from '../../api/locationServices';

export const fetchLocationsFromText = createAsyncThunk(
  'location/fetchLocationsFromText',
  async searchText => {
    const { data } = await getLocationsFromText(searchText);
    return data;
  }
);
