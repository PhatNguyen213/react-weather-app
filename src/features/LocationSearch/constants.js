export const API_URLS = {
  CITY_SEARCH: '/search.json',
};
