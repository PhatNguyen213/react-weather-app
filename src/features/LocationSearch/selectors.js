import sortBy from 'lodash/sortBy';
import { createSelector } from '@reduxjs/toolkit';

const locationState = state => state.location;

export const selectSuggestedLocations = createSelector(
  locationState,
  state => state.data
);

export const selectCityWithHighestRank = createSelector(
  selectSuggestedLocations,
  state => {
    const sorted = sortBy(state, ({ Rank }) => Rank);
    return sorted[sorted.length - 1];
  }
);

export const selectIsFetching = createSelector(
  locationState,
  state => state.loading
);
