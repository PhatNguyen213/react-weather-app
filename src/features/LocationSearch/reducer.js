import { isAnyOf } from '@reduxjs/toolkit';
import { createReducer } from '@reduxjs/toolkit';
import { fetchLocationsFromText } from './actions';

const initLocationState = {
  data: [],
  loading: false,
};

export default createReducer(initLocationState, builder => {
  builder.addCase(fetchLocationsFromText.pending, (state, { payload }) => {
    state.loading = true;
  });
  builder.addCase(fetchLocationsFromText.fulfilled, (state, { payload }) => {
    state.data = payload;
  });
  builder.addMatcher(
    isAnyOf(fetchLocationsFromText.fulfilled, fetchLocationsFromText.rejected),
    (state, { payload }) => {
      state.loading = false;
    }
  );
});
