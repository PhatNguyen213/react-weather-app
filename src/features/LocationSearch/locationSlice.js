import { API_URLS } from './constants';

const { CITY_SEARCH } = API_URLS;

export const locationSlice = {
  endpoints: builder => ({
    searchCity: builder.query({
      query: searchTerm => ({
        url: CITY_SEARCH,
        params: {
          q: searchTerm,
        },
      }),
    }),
  }),
};
