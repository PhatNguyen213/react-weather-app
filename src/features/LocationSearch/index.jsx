import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { selectSuggestedLocations } from './selectors';
import { fetchLocationsFromText } from './actions';
import LocationSearch from './LocationSearch';

const LocationSearchContainer = ({ searchTerm, onChange, onSelect }) => {
  const suggestions = useSelector(selectSuggestedLocations);
  const dispatch = useDispatch();

  useEffect(() => {
    if (searchTerm) dispatch(fetchLocationsFromText(searchTerm));
  }, [dispatch, searchTerm]);

  return (
    <LocationSearch
      suggestions={suggestions}
      searchTerm={searchTerm}
      onChange={onChange}
      onSelect={onSelect}
    />
  );
};

export default LocationSearchContainer;
