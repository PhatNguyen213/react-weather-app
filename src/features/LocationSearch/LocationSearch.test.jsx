import { getByText } from '@testing-library/react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import snapshotDiff from 'snapshot-diff';
import '@testing-library/jest-dom';

import LocationSearch from './LocationSearch';

describe('LocationSearch', () => {
  describe('renders correctly', () => {
    it('compare snapshots', () => {
      const { asFragment } = render(
        <LocationSearch
          suggestions={[
            { name: 'Test Suggestion' },
            { name: 'Test Suggestion1' },
          ]}
          searchTerm="test"
          onChange={jest.fn()}
        />
      );
      const firstRender = asFragment();
      userEvent.type(screen.getByRole('textbox'), 'test');
      expect(snapshotDiff(firstRender, asFragment())).toMatchSnapshot();
    });
  });
  describe('searchTerm prop', () => {
    it('shows default placeholder when searchTerm is tempy', () => {
      render(<LocationSearch searchTerm="" />);
      expect(screen.getByPlaceholderText('Enter a city')).toBeInTheDocument();
      expect(screen.getByRole('textbox')).toHaveValue('');
    });
    it('shows matching value when searchTerm is non-tempy', () => {
      render(<LocationSearch searchTerm="search something" />);
      expect(screen.getByRole('textbox')).toHaveValue('search something');
    });
  });
  describe('onChange prop', () => {
    it('onChange is called with user input', () => {
      const mockFn = jest.fn();
      render(<LocationSearch onChange={mockFn} searchTerm="" />);
      const input = screen.getByRole('textbox');
      userEvent.type(input, 'a');
      expect(mockFn).toBeCalledWith('a');
    });
  });
  describe('suggestions prop', () => {
    it('suggestions are shown when prop is non-tempy', () => {
      render(
        <LocationSearch
          suggestions={[
            { name: 'Test Suggestion' },
            { name: 'Test Suggestion1' },
          ]}
          searchTerm="test"
          onChange={jest.fn()}
        />
      );
      const input = screen.getByRole('textbox');
      userEvent.type(input, 'test');
      expect(screen.getByText('Test Suggestion')).toBeInTheDocument();
    });
  });
  describe('onSelect prop', () => {
    it('onSelect is called when user selects a suggestion', () => {
      const mockFn = jest.fn();
      render(
        <LocationSearch
          suggestions={[
            { name: 'Test Suggestion' },
            { name: 'Test Suggestion1' },
          ]}
          searchTerm="test"
          onChange={jest.fn()}
          onSelect={mockFn}
        />
      );
      const input = screen.getByRole('textbox');
      userEvent.type(input, 'test');
      userEvent.click(screen.getByText('Test Suggestion'));
      expect(mockFn).toBeCalledTimes(1);
    });
  });
});
