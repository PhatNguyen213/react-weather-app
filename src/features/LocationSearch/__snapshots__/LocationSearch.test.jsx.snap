// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSearch renders correctly compare snapshots 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -3,26 +3,55 @@
      class=\\"css-jqwuv0\\"
    >
      Please enter city name below for suggestions
    </div>
    <div
-     aria-expanded=\\"false\\"
+     aria-expanded=\\"true\\"
      aria-haspopup=\\"listbox\\"
      aria-owns=\\"react-autowhatever-1\\"
      class=\\"w-fit relative\\"
      role=\\"combobox\\"
    >
      <input
        aria-autocomplete=\\"list\\"
        aria-controls=\\"react-autowhatever-1\\"
        autocomplete=\\"off\\"
-       class=\\"font-sans text-base font-light w-60 h-12 px-2.5 py-5 border-solid border border-gray-600 rounded\\"
+       class=\\"font-sans text-base font-light w-60 h-12 px-2.5 py-5 border-solid border border-gray-600 rounded outline-none\\"
        placeholder=\\"Enter a city\\"
        type=\\"text\\"
        value=\\"test\\"
      />
      <div
+       class=\\"font-sans text-base font-light bg-white absolute top-51px w-72 block border-solid border border-gray-600 z-10\\"
        id=\\"react-autowhatever-1\\"
        role=\\"listbox\\"
-     />
+     >
+       <ul
+         class=\\"list-none m-0 p-0\\"
+         role=\\"listbox\\"
+       >
+         <li
+           aria-selected=\\"false\\"
+           class=\\"px-2.5 py-5 cursor-pointer\\"
+           data-suggestion-index=\\"0\\"
+           id=\\"react-autowhatever-1--item-0\\"
+           role=\\"option\\"
+         >
+           <div>
+             Test Suggestion
+           </div>
+         </li>
+         <li
+           aria-selected=\\"false\\"
+           class=\\"px-2.5 py-5 cursor-pointer\\"
+           data-suggestion-index=\\"1\\"
+           id=\\"react-autowhatever-1--item-1\\"
+           role=\\"option\\"
+         >
+           <div>
+             Test Suggestion1
+           </div>
+         </li>
+       </ul>
+     </div>
    </div>
  </DocumentFragment>"
`;
